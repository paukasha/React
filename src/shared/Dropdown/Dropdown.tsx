import React from 'react';
import styles from './dropdown.scss';
import set = Reflect.set;

interface IDropdownProps {
  // создаем кнопку на которую будем нажимать
  button: React.ReactNode;
  // это то что будет выпадать из списка
  children: React.ReactNode;
  isOpen?: boolean;
  // колбэки
  onOpen?: () => void;
  onClose?: () => void;

}
const NOOP = () => {};
export function Dropdown({ button, children, isOpen, onOpen = NOOP, onClose = NOOP }: IDropdownProps) {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState(isOpen)
  // менять начальное состояние поможет useefeect
  React.useEffect(() => {
    setIsDropdownOpen(isOpen)
  }, [isOpen])

  // также нам надо активировать колбэки onopen и onclose
  React.useEffect(() => isDropdownOpen ? onOpen() : onClose(),[isDropdownOpen])

  // введем собственный handler который будет зависеть от свойства isopen
  //будет хэндлить закрытие открытие списка

  const handleOpen = () => {
    // tclb мы передали переменную изопен
    if (isOpen === undefined) {
    //  если мы передали undefined то есть ничего то мы хотим чтобы список рботал переключал состояниеавтоматичеси
      setIsDropdownOpen(!isDropdownOpen)
    }
  }
  // сделаем дропдауну возможнмть котролитровать свое поведение сверху
  return (
  <div className={styles.container}>
  <div onClick={() => setIsDropdownOpen(!isDropdownOpen)}>
    { button }
  </div>
    {isDropdownOpen && (
      <div className={styles.listContainer}>
        {/* по нажати. на любой элемент в чилдрене*/}
        <div className={styles.list}
          onClick={() => setIsDropdownOpen(false)}>
          {children}
        </div>
      </div>
    )}
  </div>
  );
}

// const [isDropdownOpen, setIsDropdownOpen] = React.useState(isOpen);
// React.useEffect(() => setIsDropdownOpen(isOpen),[isOpen]);
// React.useEffect(() => isDropdownOpen ? onOpen(): onClose(), [ isDropdownOpen])
// const handleOpen = () => {
//   if (isOpen === undefined) {
//     setIsDropdownOpen(!isDropdownOpen)
//   }
// }

// <div className={styles.container}>
//   <div onClick={handleOpen}>
//     {button}
//   </div>
//   {isDropdownOpen && (
//     <div className={styles.listContainer}>
//       <div className={styles.list} onClick={() => setIsDropdownOpen(false)}>
//         {children}
//       </div>
//     </div>
//   )}
// </div>
