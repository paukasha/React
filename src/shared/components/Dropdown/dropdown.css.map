{"version":3,"sourceRoot":"","sources":["dropdown.scss"],"names":[],"mappings":"AAAA;EACE;EACA;;;AAGF;EACE;;;AAGF;EACE;EACA;EACA;EACA;EACA","file":"dropdown.css"}