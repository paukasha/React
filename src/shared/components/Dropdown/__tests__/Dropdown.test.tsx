/**
 * @jest-environment jsdom
 */
import React from 'react'
import {shallow} from "enzyme";
import {Dropdown} from "../Dropdown";

describe('dropdown', () => {
  test('should render', () => {
    const wrapper = shallow(<Dropdown children={<div />} button={<button />} />)
    //  проверка на то что мы отрендирили компонент
    expect(wrapper).toBeDefined();
  //   проверим зарендерилась ли кнопка
    expect(wrapper.find('div.container').isEmptyRender()).toBeFalsy();
  })

  test('should render (snapshot)', () => {
    const wrapper = shallow(<Dropdown children={<div />} button={<button />} />)
    expect(wrapper).toMatchSnapshot()
  })
})
