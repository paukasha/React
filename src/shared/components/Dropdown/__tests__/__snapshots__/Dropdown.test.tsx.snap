// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown should render (snapshot) 1`] = `
<div
  className="container"
>
  <div
    onClick={[Function]}
  >
    <button />
  </div>
</div>
`;
